## K-means on the Full Feature Sampled Dataset

#### Load Libraries

In [15]:
import pandas as pd
import os
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score
from scipy.stats import mode
import numpy as np

os.chdir(r"C:\Users\ameli\Documents\GitHub\CS4630_Group3_Project2")

#### Read Data In

In [12]:
X = pd.read_csv("data/processed/X_sample.csv")
print("X Shape:", X.shape)
print("\nFirst rows of X:")
print(X.head())

y = pd.read_csv("data/processed/y_sample.csv")
print("y Shape:", y.shape)
print("\nFirst rows of y:")
print(y.head())

X Shape: (200000, 28)

First rows of X:
   lepton_pT  lepton_eta  lepton_phi  missing_E_mag  missing_E_phi   jet1_pT  \
0   0.548478   -0.058492    0.979747       0.886419       0.773973  1.551918   
1   0.577577   -0.320490    0.009849       0.824704      -1.251037  0.706566   
2   5.065509   -1.502888   -0.564538       1.623411      -0.701304  1.297067   
3   0.851907   -0.946752    0.742821       1.373478       1.176896  0.705192   
4   0.520844    0.889179   -1.101643       0.585859      -0.184083  1.157915   

   jet1_eta  jet1_phi  jet1_btag   jet2_pT  ...  jet4_eta  jet4_phi  \
0  0.937722  0.576157   2.173076  1.391360  ...  2.452204 -1.606737   
1 -1.638856 -0.636922   0.000000  1.251607  ... -1.203890  1.031420   
2 -2.412226  0.922641   0.000000  2.297109  ... -1.362127 -1.156149   
3  0.044535 -0.453421   2.173076  0.585955  ... -1.152255 -1.049051   
4  0.453500  0.072784   0.000000  0.910912  ...  1.025577 -1.585096   

   jet4_btag      m_jj     m_jjj      m_lv     m_jlv

#### Scale Features

In [11]:
scaler = StandardScaler()

X_scaled = scaler.fit_transform(X)

In [13]:
kmeans = KMeans(n_clusters=2, random_state = 42, n_init='auto')

kmeans.fit(X_scaled)

y_kmeans = kmeans.predict(X_scaled)

print(y_kmeans)
print(y)

# labels = kmeans.labels_
# centroids = kmeans.cluster_centers_
# print(labels)
# print(centroids)

[0 1 0 ... 0 1 1]
        label
0         0.0
1         1.0
2         1.0
3         1.0
4         0.0
...       ...
199995    1.0
199996    0.0
199997    1.0
199998    0.0
199999    1.0

[200000 rows x 1 columns]


#### Evaluate with y labels

Since K-means is unsupervised, the clusters might not correspond with the true label. Since we have the label, the following code below maps the cluster labels to the majority true label inside each cluster. 

In [ ]:
mapped_labels = np.zeros_like(y_kmeans) # same length as y_kmeans

for i in range(2):
    bool_mask = (y_kmeans==i)
    mapped_labels[bool_mask] = mode(y[bool_mask], keepdims=True).mode[0]


print(f"The accuracy of the K-means algorithm on the full feature sampled dataset is {accuracy_score(y, mapped_labels)*100}%.")

The accuracy of the K-means algorithm on the full feature sampled dataset is 55.649499999999996%.


The K-means clustering achieved an accuracy of ~ 55.65% when the cluster labels were mapped to the majority class within each cluster. This is slightly higher than the baseline accuracy of 54.33%, which was calculated in the `prepare_data.ipynb`. This indicates the clusters that the K-means algorithm formed do not align well with true signal/background labels. This makes sense, as K-means groups observations based on feature similarity rather than class labels. The full feature dataset contains a lot of similar features that overlap, so the accuracy should be improved after PCA. 